In [ ]:
# Initialize Otter
import otter
grader = otter.Notebook("hw8.ipynb")

# CPSC 330 - Applied Machine Learning

## Homework 8: Introduction to Computer vision and Time Series

**Due date: see the [Calendar](https://htmlpreview.github.io/?https://github.com/UBC-CS/cpsc330/blob/master/docs/calendar.html).**

## Imports

In [1]:
from hashlib import sha1

import numpy as np
import pandas as pd
import matplotlib.pyplot as plt

from sklearn.impute import SimpleImputer
from sklearn.compose import ColumnTransformer
from sklearn.pipeline import Pipeline
from sklearn.preprocessing import StandardScaler, OrdinalEncoder, OneHotEncoder

from sklearn.linear_model import Ridge
from sklearn.ensemble import RandomForestRegressor

from sklearn.metrics import r2_score

<!-- BEGIN QUESTION -->

<div class="alert alert-info">
    
## Instructions
rubric={points}

You will earn points for following these instructions and successfully submitting your work on Gradescope.  

### Group wotk instructions

**You may work with a partner on this homework and submit your assignment as a group.** Below are some instructions on working as a group.  
- The maximum group size is 2.
  
- Use group work as an opportunity to collaborate and learn new things from each other.
- Be respectful to each other and make sure you understand all the concepts in the assignment well.
- It's your responsibility to make sure that the assignment is submitted by one of the group members before the deadline.
- You can find the instructions on how to do group submission on Gradescope [here](https://help.gradescope.com/article/m5qz2xsnjy-student-add-group-members).
- If you would like to use late tokens for the homework, all group members must have the necessary late tokens available. Please note that the late tokens will be counted for all members of the group.   


### General submission instructions

- Please **read carefully
[Use of Generative AI policy](https://ubc-cs.github.io/cpsc330-2025W1/syllabus.html#use-of-generative-ai-in-the-course)** before starting the homework assignment.
- **Run all cells before submitting:** Go to `Kernel -> Restart Kernel and Clear All Outputs`, then select `Run -> Run All Cells`. This ensures your notebook runs cleanly from start to finish without errors.
  
- **Submit your files on Gradescope.**  
   - Upload only your `.ipynb` file **with outputs displayed** and any required output files.
     
   - Do **not** submit other files from your repository.  
   - If you need help, see the [Gradescope Student Guide](https://lthub.ubc.ca/guides/gradescope-student-guide/).  
- **Check that outputs render properly.**  
   - Make sure all plots and outputs appear in your submission.
     
   - If your `.ipynb` file is too large and doesn't render on Gradescope, also upload a PDF or HTML version so the TAs can view your work.  
- **Keep execution order clean.**  
   - Execution numbers must start at "1" and increase in order.
     
   - Notebooks without visible outputs may not be graded.  
   - Out-of-order or missing execution numbers may result in mark deductions.  
- **Follow course submission guidelines:** Review the [CPSC 330 homework instructions](https://ubc-cs.github.io/cpsc330-2025W1/docs/homework_instructions.html) for detailed guidance on completing and submitting assignments.
   
</div>

_Points:_ 2

<!-- END QUESTION -->

<br><br>

## Exercise 1: time series prediction

In this exercise we'll be looking at a [dataset of avocado prices](https://www.kaggle.com/neuromusic/avocado-prices). You should start by downloading the dataset and storing it under the `data` folder. We will be forcasting average avocado price for the next week.

In [3]:
df = pd.read_csv("data/avocado.csv", parse_dates=["Date"], index_col=0)
df.head()

Date  AveragePrice  Total Volume     4046       4225    4770  \
0 2015-12-27          1.33      64236.62  1036.74   54454.85   48.16   
1 2015-12-20          1.35      54876.98   674.28   44638.81   58.33   
2 2015-12-13          0.93     118220.22   794.70  109149.67  130.50   
3 2015-12-06          1.08      78992.15  1132.00   71976.41   72.58   
4 2015-11-29          1.28      51039.60   941.48   43838.39   75.78   

   Total Bags  Small Bags  Large Bags  XLarge Bags          type  year  region  
0     8696.87     8603.62       93.25          0.0  conventional  2015  Albany  
1     9505.56     9408.07       97.49          0.0  conventional  2015  Albany  
2     8145.35     8042.21      103.14          0.0  conventional  2015  Albany  
3     5811.16     5677.40      133.76          0.0  conventional  2015  Albany  
4     6183.95     5986.26      197.69          0.0  conventional  2015  Albany

In [4]:
df.shape

(18249, 13)

In [5]:
df["Date"].min()

Timestamp('2015-01-04 00:00:00')

In [6]:
df["Date"].max()

Timestamp('2018-03-25 00:00:00')

It looks like the data ranges from the start of 2015 to March 2018 (~2 years ago), for a total of 3.25 years or so. Let's split the data so that we have a 6 months of test data.

In [7]:
split_date = '20170925'
df_train = df[df["Date"] <= split_date]
df_test  = df[df["Date"] >  split_date]

In [8]:
assert len(df_train) + len(df_test) == len(df)

<br><br>

<!-- BEGIN QUESTION -->

### 1.1 How many time series?
rubric={points:4}

In the [Rain in Australia](https://www.kaggle.com/datasets/jsphyg/weather-dataset-rattle-package) dataset from lecture demo, we had different measurements for each Location.

We want you to consider this for the avocado prices dataset. For which categorical feature(s), if any, do we have separate measurements? Justify your answer by referencing the dataset.

<div class="alert alert-warning">

Solution_1.1
    
</div>

_Points:_ 4

We have separate measurements for the categorical features region and type.
For any given date, the dataset contains multiple rows that differ by region, and within each region the prices are also split into conventional vs. organic avocados. This means the dataset is composed of multiple parallel time series, meaning one time series for each (region × type) combination.

In [9]:
#Unique features
df["region"].nunique(), df["type"].nunique()

(54, 2)

In [10]:
#total time series
df.groupby(["region", "type"]).ngroups

108

In [11]:
df.groupby(["region", "type"]).size().head()

region               type        
Albany               conventional    169
                     organic         169
Atlanta              conventional    169
                     organic         169
BaltimoreWashington  conventional    169
dtype: int64

There are 108 different time series, where a series exists for each group & type combination

<!-- END QUESTION -->

<br><br>

<!-- BEGIN QUESTION -->

### 1.2 Equally spaced measurements?
rubric={points:4}

In the Rain in Australia dataset, the measurements were generally equally spaced but with some exceptions. How about with this dataset? Justify your answer by referencing the dataset.

<div class="alert alert-warning">

Solution_1.2
    
</div>

_Points:_ 4

The measurements in this dataset are not perfectly equally spaced. When I compute the time differences between consecutive dates within each (region, type) time series, most gaps are 7 days (weekly data), but I also see some gaps of 14 and 21 days. This means that for some dates, measurements are missing, causing longer intervals. Therefore, unlike the mostly-regular Rain in Australia dataset, the avocado dataset contains irregularly spaced time series with occasional missing weeks.

In [12]:
# Sort so that dates within each time series are in order
df_sorted = df.sort_values(["region", "type", "Date"])
df_sorted.head()

Date  AveragePrice  Total Volume     4046      4225    4770  \
51 2015-01-04          1.22      40873.28  2819.50  28287.42   49.90   
50 2015-01-11          1.24      41195.08  1002.85  31640.34  127.12   
49 2015-01-18          1.17      44511.28   914.14  31540.32  135.77   
48 2015-01-25          1.06      45147.50   941.38  33196.16  164.14   
47 2015-02-01          0.99      70873.60  1353.90  60017.20  179.32   

    Total Bags  Small Bags  Large Bags  XLarge Bags          type  year  \
51     9716.46     9186.93      529.53          0.0  conventional  2015   
50     8424.77     8036.04      388.73          0.0  conventional  2015   
49    11921.05    11651.09      269.96          0.0  conventional  2015   
48    10845.82    10103.35      742.47          0.0  conventional  2015   
47     9323.18     9170.82      152.36          0.0  conventional  2015   

    region  
51  Albany  
50  Albany  
49  Albany  
48  Albany  
47  Albany

In [13]:
# Difference in days between consecutive dates for each (region, type)
deltas = df_sorted.groupby(["region", "type"])["Date"].diff().dt.days
deltas.head()

51    NaN
50    7.0
49    7.0
48    7.0
47    7.0
Name: Date, dtype: float64

In [14]:
# Difference in days between consecutive dates for each (region, type)
deltas = df_sorted.groupby(["region", "type"])["Date"].diff().dt.days
deltas.head()

51    NaN
50    7.0
49    7.0
48    7.0
47    7.0
Name: Date, dtype: float64

In [15]:
# all unique time gaps
deltas_unique = deltas.dropna().unique()
deltas_unique

array([ 7., 14., 21.])

In [16]:
# Minimum and maximum time gaps
deltas.min(), deltas.max()

(7.0, 21.0)

In [17]:
# check if there are time gaps different from 7 days
(deltas.dropna() != 7).any()

np.True_

In [18]:
# Rows where the gap is not 7 days (if any)
df_sorted.loc[deltas != 7].head()

Date  AveragePrice  Total Volume       4046       4225      4770  \
51 2015-01-04          1.22      40873.28    2819.50   28287.42     49.90   
51 2015-01-04          1.79       1373.95      57.42     153.88      0.00   
51 2015-01-04          1.00     435021.49  364302.39   23821.16     82.15   
51 2015-01-04          1.76       3846.69    1500.15     938.35      0.00   
51 2015-01-04          1.08     788025.06   53987.31  552906.04  39995.03   

    Total Bags  Small Bags  Large Bags  XLarge Bags          type  year  \
51     9716.46     9186.93      529.53          0.0  conventional  2015   
51     1162.65     1162.65        0.00          0.0       organic  2015   
51    46815.79    16707.15    30108.64          0.0  conventional  2015   
51     1408.19     1071.35      336.84          0.0       organic  2015   
51   141136.68   137146.07     3990.61          0.0  conventional  2015   

                 region  
51               Albany  
51               Albany  
51              Atlanta  
51              Atlanta  
51  BaltimoreWashington

<!-- END QUESTION -->

<br><br>

<!-- BEGIN QUESTION -->

### 1.3 Interpreting regions
rubric={points:4}

In the Rain in Australia dataset, each location was a different place in Australia. For this dataset, look at the names of the regions. Do you think the regions are also all distinct, or are there overlapping regions? Justify your answer by referencing the data.

<div class="alert alert-warning">

Solution_1.3
    
</div>

_Points:_ 4

The regions in this dataset are not all clearly distinct. Some region names represent specific cities (e.g., "Albany", "Chicago"), while others correspond to much larger, overlapping areas such as "California", "Northeast", "West", or "TotalUS". These larger regions include or overlap with many of the smaller city-level regions. Because of this, the regions are not mutually exclusive geographic locations, unlike the Rain in Australia dataset. Instead, the avocado dataset mixes local markets and aggregated multi-state markets, so many regions overlap rather than being distinct.

In [19]:
df["region"].unique()

array(['Albany', 'Atlanta', 'BaltimoreWashington', 'Boise', 'Boston',
       'BuffaloRochester', 'California', 'Charlotte', 'Chicago',
       'CincinnatiDayton', 'Columbus', 'DallasFtWorth', 'Denver',
       'Detroit', 'GrandRapids', 'GreatLakes', 'HarrisburgScranton',
       'HartfordSpringfield', 'Houston', 'Indianapolis', 'Jacksonville',
       'LasVegas', 'LosAngeles', 'Louisville', 'MiamiFtLauderdale',
       'Midsouth', 'Nashville', 'NewOrleansMobile', 'NewYork',
       'Northeast', 'NorthernNewEngland', 'Orlando', 'Philadelphia',
       'PhoenixTucson', 'Pittsburgh', 'Plains', 'Portland',
       'RaleighGreensboro', 'RichmondNorfolk', 'Roanoke', 'Sacramento',
       'SanDiego', 'SanFrancisco', 'Seattle', 'SouthCarolina',
       'SouthCentral', 'Southeast', 'Spokane', 'StLouis', 'Syracuse',
       'Tampa', 'TotalUS', 'West', 'WestTexNewMexico'], dtype=object)

In [20]:
sorted(df["region"].unique())

['Albany',
 'Atlanta',
 'BaltimoreWashington',
 'Boise',
 'Boston',
 'BuffaloRochester',
 'California',
 'Charlotte',
 'Chicago',
 'CincinnatiDayton',
 'Columbus',
 'DallasFtWorth',
 'Denver',
 'Detroit',
 'GrandRapids',
 'GreatLakes',
 'HarrisburgScranton',
 'HartfordSpringfield',
 'Houston',
 'Indianapolis',
 'Jacksonville',
 'LasVegas',
 'LosAngeles',
 'Louisville',
 'MiamiFtLauderdale',
 'Midsouth',
 'Nashville',
 'NewOrleansMobile',
 'NewYork',
 'Northeast',
 'NorthernNewEngland',
 'Orlando',
 'Philadelphia',
 'PhoenixTucson',
 'Pittsburgh',
 'Plains',
 'Portland',
 'RaleighGreensboro',
 'RichmondNorfolk',
 'Roanoke',
 'Sacramento',
 'SanDiego',
 'SanFrancisco',
 'Seattle',
 'SouthCarolina',
 'SouthCentral',
 'Southeast',
 'Spokane',
 'StLouis',
 'Syracuse',
 'Tampa',
 'TotalUS',
 'West',
 'WestTexNewMexico']

In [21]:
df["region"].nunique()

54

<!-- END QUESTION -->

<br><br>

We will use the entire dataset despite any location-based weirdness uncovered in the previous part.

We will be trying to forecast the avocado price. The function below is adapted from the lecture.

In [22]:
import pandas as pd


def create_lag_feature(
    df: pd.DataFrame,
    orig_feature: str,
    lag: int,
    groupby: list[str],
    new_feature_name: str | None = None,
    clip: bool = False,
) -> pd.DataFrame:
    """
    Create a lagged (or ahead) version of a feature, optionally per group.

    Assumes df is already sorted by time within each group and has unique indices.

    Parameters
    ----------
    df : pd.DataFrame
        The dataset.
    orig_feature : str
        Name of the column to lag.
    lag : int
        The lag:
          - negative → values from the past (t-1, t-2, ...)
          - positive → values from the future (t+1, t+2, ...)
    groupby : list of str
        Column(s) to group by if df contains multiple time series.
    new_feature_name : str, optional
        Name of the new column. If None, a name is generated automatically.
    clip : bool, default False
        If True, drop rows where the new feature is NaN.

    Returns
    -------
    pd.DataFrame
        A new dataframe with the additional column added.
    """
    if lag == 0:
        raise ValueError("lag cannot be 0 (no shift). Use the original feature instead.")

    # Default name if not provided
    if new_feature_name is None:
        if lag < 0:
            new_feature_name = f"{orig_feature}_lag{abs(lag)}"
        else:
            new_feature_name = f"{orig_feature}_ahead{lag}"

    df = df.copy()

    # Map your convention (negative=past, positive=future) to pandas shift
    # pandas: shift(+k) → past, shift(-k) → future
    periods = abs(lag) if lag < 0 else -lag

    df[new_feature_name] = (
        df.groupby(groupby, sort=False)[orig_feature]
          .shift(periods)
    )

    if clip:
        df = df.dropna(subset=[new_feature_name])

    return df


We first sort our dataframe properly:

In [23]:
df_sort = df.sort_values(by=["region", "type", "Date"]).reset_index(drop=True)
df_sort

Date  AveragePrice  Total Volume     4046      4225    4770  \
0     2015-01-04          1.22      40873.28  2819.50  28287.42   49.90   
1     2015-01-11          1.24      41195.08  1002.85  31640.34  127.12   
2     2015-01-18          1.17      44511.28   914.14  31540.32  135.77   
3     2015-01-25          1.06      45147.50   941.38  33196.16  164.14   
4     2015-02-01          0.99      70873.60  1353.90  60017.20  179.32   
...          ...           ...           ...      ...       ...     ...   
18244 2018-02-25          1.57      18421.24  1974.26   2482.65    0.00   
18245 2018-03-04          1.54      17393.30  1832.24   1905.57    0.00   
18246 2018-03-11          1.56      22128.42  2162.67   3194.25    8.93   
18247 2018-03-18          1.56      15896.38  2055.35   1499.55    0.00   
18248 2018-03-25          1.62      15303.40  2325.30   2171.66    0.00   

       Total Bags  Small Bags  Large Bags  XLarge Bags          type  year  \
0         9716.46     9186.93      529.53          0.0  conventional  2015   
1         8424.77     8036.04      388.73          0.0  conventional  2015   
2        11921.05    11651.09      269.96          0.0  conventional  2015   
3        10845.82    10103.35      742.47          0.0  conventional  2015   
4         9323.18     9170.82      152.36          0.0  conventional  2015   
...           ...         ...         ...          ...           ...   ...   
18244    13964.33    13698.27      266.06          0.0       organic  2018   
18245    13655.49    13401.93      253.56          0.0       organic  2018   
18246    16762.57    16510.32      252.25          0.0       organic  2018   
18247    12341.48    12114.81      226.67          0.0       organic  2018   
18248    10806.44    10569.80      236.64          0.0       organic  2018   

                 region  
0                Albany  
1                Albany  
2                Albany  
3                Albany  
4                Albany  
...                 ...  
18244  WestTexNewMexico  
18245  WestTexNewMexico  
18246  WestTexNewMexico  
18247  WestTexNewMexico  
18248  WestTexNewMexico  

[18249 rows x 13 columns]

We then call `create_lag_feature`. This creates a new column in the dataset `AveragePriceNextWeek`, which is the following week's `AveragePrice`. We have set `clip=True` which means it will remove rows where the target would be missing.

In [24]:
df_hastarget = create_lag_feature(df_sort, "AveragePrice", +1, ["region", "type"], "AveragePriceNextWeek", clip=True)
df_hastarget

Date  AveragePrice  Total Volume     4046      4225    4770  \
0     2015-01-04          1.22      40873.28  2819.50  28287.42   49.90   
1     2015-01-11          1.24      41195.08  1002.85  31640.34  127.12   
2     2015-01-18          1.17      44511.28   914.14  31540.32  135.77   
3     2015-01-25          1.06      45147.50   941.38  33196.16  164.14   
4     2015-02-01          0.99      70873.60  1353.90  60017.20  179.32   
...          ...           ...           ...      ...       ...     ...   
18243 2018-02-18          1.56      17597.12  1892.05   1928.36    0.00   
18244 2018-02-25          1.57      18421.24  1974.26   2482.65    0.00   
18245 2018-03-04          1.54      17393.30  1832.24   1905.57    0.00   
18246 2018-03-11          1.56      22128.42  2162.67   3194.25    8.93   
18247 2018-03-18          1.56      15896.38  2055.35   1499.55    0.00   

       Total Bags  Small Bags  Large Bags  XLarge Bags          type  year  \
0         9716.46     9186.93      529.53          0.0  conventional  2015   
1         8424.77     8036.04      388.73          0.0  conventional  2015   
2        11921.05    11651.09      269.96          0.0  conventional  2015   
3        10845.82    10103.35      742.47          0.0  conventional  2015   
4         9323.18     9170.82      152.36          0.0  conventional  2015   
...           ...         ...         ...          ...           ...   ...   
18243    13776.71    13553.53      223.18          0.0       organic  2018   
18244    13964.33    13698.27      266.06          0.0       organic  2018   
18245    13655.49    13401.93      253.56          0.0       organic  2018   
18246    16762.57    16510.32      252.25          0.0       organic  2018   
18247    12341.48    12114.81      226.67          0.0       organic  2018   

                 region  AveragePriceNextWeek  
0                Albany                  1.24  
1                Albany                  1.17  
2                Albany                  1.06  
3                Albany                  0.99  
4                Albany                  0.99  
...                 ...                   ...  
18243  WestTexNewMexico                  1.57  
18244  WestTexNewMexico                  1.54  
18245  WestTexNewMexico                  1.56  
18246  WestTexNewMexico                  1.56  
18247  WestTexNewMexico                  1.62  

[18141 rows x 14 columns]

Our goal is to predict `AveragePriceNextWeek`.

Let's split the data:

In [25]:
df_train = df_hastarget[df_hastarget["Date"] <= split_date]
df_test  = df_hastarget[df_hastarget["Date"] >  split_date]

<br><br>

<!-- BEGIN QUESTION -->

### 1.4 `AveragePrice` baseline
rubric={points}

Soon we will want to build some models to forecast the average avocado price a week in advance. Before we start with any ML though, let's try a baseline. Previously we used `DummyClassifier` or `DummyRegressor` as a baseline. This time, we'll do something else as a baseline: we'll assume the price stays the same from this week to next week. So, we'll set our prediction of "AveragePriceNextWeek" exactly equal to "AveragePrice", assuming no change. That is kind of like saying, "If it's raining today then I'm guessing it will be raining tomorrow". This simplistic approach will not get a great score but it's a good starting point for reference. If our model does worse that this, it must not be very good.

Using this baseline approach, what $R^2$ do you get on the train and test data?

<div class="alert alert-warning">

Solution_1.4
    
</div>

_Points:_ 4

Using the “no-change” baseline, where we predict AveragePriceNextWeek to be equal to this week’s AveragePrice, I get
R² ≈ 0.83 on the training data and R² ≈ 0.76 on the test data. So this simple baseline already explains a good amount of the variance, and any ML model we build later should aim to beat these scores.

In [26]:
train_r2 = None
y_train_true = df_train["AveragePriceNextWeek"]
y_train_pred = df_train["AveragePrice"]   # baseline prediction for train set

In [27]:
test_r2 = None

y_test_true = df_test["AveragePriceNextWeek"]
y_test_pred = df_test["AveragePrice"]    # baseline prediction for test set

In [28]:
train_r2 = r2_score(y_train_true, y_train_pred)
train_r2

0.8285800937261841

In [29]:
test_r2 = r2_score(y_test_true, y_test_pred)
test_r2

0.7631780188583048

In [30]:
assert not train_r2 is None, "Are you using the correct variable name?"
assert not test_r2 is None, "Are you using the correct variable name?"
assert sha1(str(round(train_r2, 3)).encode('utf8')).hexdigest() == 'b1136fe2a8918904393ab6f40bfb3f38eac5fc39', "Your training score is not correct. Are you using the right features?"
assert sha1(str(round(test_r2, 3)).encode('utf8')).hexdigest() == 'cc24d9a9b567b491a56b42f7adc582f2eefa5907', "Your test score is not correct. Are you using the right features?"

<!-- END QUESTION -->

<br><br>

<!-- BEGIN QUESTION -->

### 1.5 Forecasting average avocado price
rubric={points:10}

Now that the baseline is done, let's build some models to forecast the average avocado price a week later. Experiment with a few approachs for encoding the date. Justify the decisions you make. Which approach worked best? Report your test score and briefly discuss your results.

Benchmark: you should be able to achieve $R^2$ of at least 0.79 on the test set. I got to 0.80, but not beyond that. Let me know if you do better!

Note: because we only have 2 splits here, we need to be a bit wary of overfitting on the test set. Try not to test on it a ridiculous number of times. If you are interested in some proper ways of dealing with this, see for example sklearn's [TimeSeriesSplit](https://scikit-learn.org/stable/modules/generated/sklearn.model_selection.TimeSeriesSplit.html), which is like cross-validation for time series data.

<div class="alert alert-warning">

Solution_1.5
    
</div>

_Points:_ 10

To forecast AveragePriceNextWeek, I experimented with building a full preprocessing + modeling pipeline. I started from the baseline assumption that next week's price equals this week's price, which gave a test R² of about 0.76. The goal was to improve on this baseline by adding useful features and trying a more flexible model. I engineered simple date features from the Date column (year, month, weekofyear) using a FunctionTransformer, since raw timestamps are not directly informative for forests/tree ensemble models.

I also applied one-hot encoding to the categorical variables (region and type) and standardized all numeric features (e.g., price, volumes, bag counts). All preprocessing was wrapped inside a ColumnTransformer and integrated into a single scikit-learn pipeline. For the model, I used XGBoostRegressor, and also performed some hyperparamtertuning with the help of TimeSeriesSplit to ensure validation sets are not shuffled around. With these settings and tuned hyperparameters for XGBoost, the model achieved:

Train R² ≈ 0.8793

Test R² ≈ 0.7924

Compared to the baseline, the test score has clearly improved, and it lands in the benchmark of 0.79 - 0.80 as mentioned in the instructions. Both training scores and test scores have shown to improve, but the gap between train and test score still exists. This could be an indication that there is some overfitting present in the model, but it's not too big of a concern because test score still improved noticeably.

Overall, the combination of engineered date features, one-hot encoding of region/type, proper numerical scaling, and a moderately hypertuned XGBoost ensemble model led to the best performance on this dataset.

In [39]:
from sklearn.model_selection import train_test_split
from sklearn.compose import ColumnTransformer
from sklearn.preprocessing import OneHotEncoder, StandardScaler, FunctionTransformer
from sklearn.pipeline import Pipeline
from sklearn.ensemble import RandomForestRegressor
from sklearn.metrics import r2_score
from sklearn.linear_model import Ridge
from sklearn.model_selection import TimeSeriesSplit, RandomizedSearchCV
from scipy.stats import loguniform
from scipy.stats import randint, uniform
from xgboost import XGBRegressor
import numpy as np
import pandas as pd

In [40]:
X_train = df_train.drop(columns=["AveragePriceNextWeek"])
y_train = df_train["AveragePriceNextWeek"]

X_test = df_test.drop(columns=["AveragePriceNextWeek"])
y_test = df_test["AveragePriceNextWeek"]

In [41]:
def add_date_features(df):
    df = df.copy()
    df["year"] = df["Date"].dt.year
    df["month"] = df["Date"].dt.month
    df["weekofyear"] = df["Date"].dt.isocalendar().week.astype(int)
    return df

date_featurizer = FunctionTransformer(add_date_features, validate=False)

In [46]:
numeric_features = [
    "AveragePrice", "Total Volume", "4046", "4225", "4770",
    "Total Bags", "Small Bags", "Large Bags", "XLarge Bags",
    "year", "month", "weekofyear",
]
categorical_features = ["region", "type"]

preprocess = ColumnTransformer(
    transformers=[
        ("num", StandardScaler(), numeric_features),
        ("cat", OneHotEncoder(handle_unknown="ignore"), categorical_features),
    ],
)

tscv = TimeSeriesSplit(n_splits=5)

param_distributions = {
    "xgb__n_estimators": randint(200, 1500),
    "xgb__max_depth": randint(2, 12),
    "xgb__learning_rate": uniform(0.01, 0.3),
    "xgb__subsample": uniform(0.5, 0.5),
    "xgb__colsample_bytree": uniform(0.5, 0.5)
}

model = Pipeline(
    steps=[
        ("date", date_featurizer),
        ("preprocess", preprocess),
        ("xgb", XGBRegressor())
    ]
)

# perform hyperparameter tuning for xgboost ensemble
rs = RandomizedSearchCV(
    estimator=model,
    param_distributions=param_distributions,
    n_iter=10,
    cv=tscv,
    random_state=123
)

rs.fit(X_train, y_train)

print("Best params:", rs.best_params_)

Best params: {'xgb__colsample_bytree': np.float64(0.7115532300622305), 'xgb__learning_rate': np.float64(0.30422925951538465), 'xgb__max_depth': 2, 'xgb__n_estimators': 313, 'xgb__subsample': np.float64(0.7404659507421805)}


In [48]:
best_xg_boost_pipeline = rs.best_estimator_

train_predictions = best_xg_boost_pipeline.predict(X_train)
test_predictions = best_xg_boost_pipeline.predict(X_test)

train_r2_score = r2_score(y_train_true, train_predictions)
test_r2_score = r2_score(y_test_true, test_predictions)

train_r2_score, test_r2_score

(0.8793518471745198, 0.792497502236223)

<!-- END QUESTION -->

<br><br><br><br>

## Exercise 2: Short answer questions

<!-- BEGIN QUESTION -->

### 2.1 Time series

rubric={points:6}

The following questions pertain to Lecture 20 on time series data:

1. Sometimes a time series has missing time points or, worse, time points that are unequally spaced in general. Give an example of a real world situation where the time series data would have unequally spaced time points.
2. In class we discussed two approaches to using temporal information: encoding the date as one or more features, and creating lagged versions of features. Which of these (one/other/both/neither) two approaches would struggle with unequally spaced time points? Briefly justify your answer.
3. When studying time series modeling, we explored several ways to encode date information as a feature for the citibike dataset. When we used time of day as a numeric feature, the Ridge model was not able to capture the periodic pattern. Why? How did we tackle this problem? Briefly explain.

<div class="alert alert-warning">

Solution_2.1
    
</div>

_Points:_ 6

1. A common example is hospital patient monitoring. Measurements like heart rate or blood pressure are taken whenever nurses check on a patient, which might be irregular (e.g., 10 minutes apart, then 45 minutes, then 2 hours). Another example is user activity on apps or websites, where events only happen when a user interacts, so timestamps are naturally uneven.

2. The approach that struggles is creating lagged versions of features.
Lag features assume that data points occur at regular intervals (e.g., weekly or daily). If the time gaps are inconsistent, “lag 1” might correspond to a 1-hour gap in one part of the dataset and a 3-day gap in another, which breaks the meaning of the lag.
Encoding the date directly as features (like year/month/day) does not rely on equal spacing, so it handles irregular intervals better.

3. When we encoded “time of day” as a plain numeric feature, Ridge interpreted it linearly (e.g., 0 → 23). But time is cyclic: 23:00 and 0:00 are actually right next to each other. This linear encoding destroys the periodic structure, so Ridge couldn’t model the repeating daily pattern.
We fixed this by converting time into sinusoidal features (sin(2πt/24) and cos(2πt/24)), which correctly represent the circular nature of time. Once we added sine/cosine features, the model could capture the periodic behavior properly.

<!-- END QUESTION -->

<br><br>

<!-- BEGIN QUESTION -->

### 2.2 Computer vision
rubric={points:6}

The following questions pertain to the lecture on multiclass classification and introduction to computer vision.

1. How many parameters (coefficients and intercepts) will `sklearn`’s `LogisticRegression()` model learn for a four-class classification problem, assuming that you have 10 features? Briefly explain your answer.
2. In Lecture 19, we briefly discussed how neural networks are sort of like `sklearn`'s pipelines, in the sense that they involve multiple sequential transformations of the data, finally resulting in the prediction. Why was this property useful when it came to transfer learning?
3. Imagine that you have a small dataset with ~1000 images containing pictures and names of 50 different Computer Science faculty members from UBC. Your goal is to develop a reasonably accurate multi-class classification model for this task. Describe which model/technique you would use and briefly justify your choice in one to three sentences.

<div class="alert alert-warning">

Solution_2.2
    
</div>

_Points:_ 6

1. For multinomial logistic regression, sklearn uses one set of parameters per class.
That means:
* Weights: 4 classes × 10 features = 40 coefficients
* Intercepts: 4 intercepts = 4
So in total, the model learns 44 parameters. (Each class gets its own linear classifier before the softmax.)

2. This property matters because transfer learning works by reusing the early layers of a pretrained network.
Those early layers act like a general feature extractor (edges, textures, shapes) that work across many image tasks. Because neural nets process data through multiple stages just like a sklearn pipeline. The “front” of the pipeline can be reused, and we only retrain the last few layers for our small, specific task. This saves data, time, and avoids overfitting.

3. I would use transfer learning with a pretrained CNN, and fine-tune only the last few layers. With only ~1000 images and 50 classes, training a CNN from scratch would massively overfit. A pretrained model already knows useful visual features, so fine-tuning it lets me get strong performance even with limited data.


<!-- END QUESTION -->

<br><br>

Before submitting your assignment, please make sure you have followed all the instructions in the Submission Instructions section at the top.

Here is a quick checklist before submitting:

- [ ] Restart kernel, clear outputs, and run all cells from top to bottom.  
- [ ] `.ipynb` file runs without errors and contains all outputs.  
- [ ] Only `.ipynb` and required output files are uploaded (no extra files).  
- [ ] Execution numbers start at **1** and are in order.  
- [ ] If `.ipynb` is too large and doesn't render on Gradescope, also upload a PDF/HTML version.  
- [ ] Reviewed the [CPSC 330 homework instructions](https://ubc-cs.github.io/cpsc330-2025W1/docs/homework_instructions.html).  

![](img/eva-well-done.png)